File: core/ed25519/python/sign.py
"""
PEAC Protocol v0.9.1
Ed25519 signer (Python)
Apache 2.0 License
"""

import base64
from nacl.signing import SigningKey

def sign(message: str, private_key_b64: str, nonce: str, timestamp: int) -> str:
    sk = SigningKey(base64.b64decode(private_key_b64))
    combined = (message + nonce + str(timestamp)).encode('utf-8')
    signature = sk.sign(combined).signature
    return base64.b64encode(signature).decode()

File: archived/core/ed25519/node/sign.js
/**
 * PEAC Protocol v0.9.1
 * Ed25519 message signer (Node.js)
 * Uses libsodium-wrappers for Ed25519 signatures.
 * Apache 2.0 License
 */

const sodium = require('libsodium-wrappers');

async function sign(message, privateKeyBase64, nonce, timestamp) {
  await sodium.ready;
  const privateKey = Buffer.from(privateKeyBase64, 'base64');
  // Combine message, nonce, timestamp for replay protection
  const combined = Buffer.concat([
    Buffer.from(message),
    Buffer.from(nonce),
    Buffer.from(String(timestamp))
  ]);
  const signature = sodium.crypto_sign_detached(combined, privateKey);
  return Buffer.from(signature).toString('base64');
}

module.exports = { sign };
